### PR TITLE
chore(debug): various debug mode improvements

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/progress.ts [hunk 1]
```diff
@@ -116,6 +122,9 @@ export class ProgressController {
       this._logger.log(apiLog, `<= ${this._apiName} failed`);
       await Promise.all(this._cleanups.splice(0).map(cleanup => runCleanup(cleanup)));
       throw e;
+    } finally {
+      this._logRecording = [];
```
Comment (dgozman) [inline src/progress.ts:126]: Well, remove this line from the catch() block now.

File: src/debug/injected/recorder.ts [hunk 2]
```diff
@@ -92,24 +92,69 @@ export class Recorder {
...
+      if (selector && selector.final)
+        return fullSelector;
+      if (targetElement === this._injectedScript.querySelector(parseSelector(fullSelector), root))
+        return fullSelector;
+    }
+    return '<selector>';
+  }
+
+  private _buildSelectorCandidate(element: Element): { final: boolean, selector: string } | null {
+    for (const attribute of ['data-testid', 'aria-label', 'data-test-id', 'data-test']) {
```
Comment (dgozman) [inline src/debug/injected/recorder.ts:113]: Why aria-label is final? I thought it's similar to text.

File: src/debug/injected/recorder.ts [hunk 3]
```diff
@@ -92,24 +92,69 @@ export class Recorder {
...
-    if (element.firstChild && element.firstChild === element.lastChild && element.firstChild.nodeType === Node.TEXT_NODE)
-      return `text="${element.textContent}"`;
-    return '<selector>';
+    const textSelector = textSelectorForElement(element);
+    if (textSelector)
+      return { final: false, selector: textSelector };
+
+    // Depreoritize id, but still use it as a last resort.
+    if (element.hasAttribute('id'))
+      return { final: true, selector: `${element.nodeName.toLocaleLowerCase()}[id=${element.getAttribute('id')}]` };
```
Comment (dgozman) [inline src/debug/injected/recorder.ts:132]: With shadow dom, id does not guarantee uniqueness anymore. Perhaps, do not mark it as final?

File: src/debug/injected/recorder.ts [hunk 4]
```diff
@@ -92,24 +92,69 @@ export class Recorder {
...
+  private _buildSelector(targetElement: Element): string {
+    const path: string[] = [];
+    const root = document.documentElement;
+    for (let element: Element | null = targetElement; element && element !== root; element = element.parentElement) {
+      const selector = this._buildSelectorCandidate(element);
+      if (selector)
+        path.unshift(selector.selector);
+
+      const fullSelector = path.join(' ');
+      if (selector && selector.final)
```
Comment (dgozman) [inline src/debug/injected/recorder.ts:104]: I am curious why don't you do the sanity check for "final" selectors?

**Review by dgozman — APPROVED** [review]

File: src/debug/injected/recorder.ts (hunk 2 above)
Comment (pavelfeldman) [reply src/debug/injected/recorder.ts:113]: Dunno, just feels like a much stronger signal :/

File: src/debug/injected/recorder.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/debug/injected/recorder.ts:132]: Nothing works for shadow yet

File: src/debug/injected/recorder.ts (hunk 4 above)
Comment (pavelfeldman) [reply src/debug/injected/recorder.ts:104]: Because it is wip!

File: src/progress.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/progress.ts:126]: Done

**Pull request merged**